File: webapp/channels/src/actions/views/post_seen.ts
// Copyright (c) 2015-present Aura, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type { PostSeenReceipts } from '@mattermost/types/posts';
import { Client4 } from 'mattermost-redux/client';
import { forceLogoutIfNecessary } from 'mattermost-redux/actions/helpers';

import type { ActionFuncAsync } from 'types/store';

export function fetchPostSeenReceipts(postId: string): ActionFuncAsync<PostSeenReceipts> {
	return async (dispatch, getState) => {
		try {
			const data = await Client4.getPostSeenReceipts(postId);
			return { data };
		} catch (error) {
			forceLogoutIfNecessary(error, dispatch, getState);
			return { error: error as Error };
		}
	};
}

File: server/public/model/post_seen.go
// Copyright (c) 2015-present Aura, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package model

type PostReaderInfo struct {
	UserId    string `json:"user_id"`
	Username  string `json:"username"`
	FirstName string `json:"first_name"`
	LastName  string `json:"last_name"`
	Nickname  string `json:"nickname"`
	ViewedAt  int64  `json:"viewed_at"`
}

type PostDeliveredInfo struct {
	UserId    string `json:"user_id"`
	Username  string `json:"username"`
	FirstName string `json:"first_name"`
	LastName  string `json:"last_name"`
	Nickname  string `json:"nickname"`
}

type PostSeenReceipts struct {
	PostId      string               `json:"post_id"`
	ChannelId   string               `json:"channel_id"`
	CreateAt    int64                `json:"create_at"`
	ReadBy      []*PostReaderInfo    `json:"read_by"`
	DeliveredTo []*PostDeliveredInfo `json:"delivered_to"`
}
